feat(feature_branch): Private Kernel Circuit

kevaundray opened the pull request:
Please provide a paragraph or two giving a summary of the change, including relevant motivation and context.

# Checklist:
Remove the checklist to signal you've completed it. Enable auto-merge if the PR is ready to merge.
- [ ] If the pull request requires a cryptography review (e.g. cryptographic algorithm implementations) I have added the 'crypto' tag.
- [ ] I have reviewed my diff in github, line by line and removed unexpected formatting changes, testing logs, or commented-out code.
- [ ] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to relevant issues (if any exist).

kevaundray committed:
add noir-private-kernel yarn member

kevaundray committed:
add initial private kernel circuit implementation

kevaundray committed:
move bugs/devex files out of noir project subfolder

sirasistant committed:
feat: Use constants_gen for constants (#2742)

Re exporting globals doesn't quite work ): so I just replaced every use
with constants gen and added a missing generator index

kevaundray committed:
chore: Add barebones executeCircuit method for init circuit (#2747)

Abi to typescript is missing, so its a bit hard to get the inputs in the
correct shape. For now, I've just zeroed out the witness map and passed
it to executeCircuit.


# Checklist:
Remove the checklist to signal you've completed it. Enable auto-merge if
the PR is ready to merge.
- [ ] If the pull request requires a cryptography review (e.g.
cryptographic algorithm implementations) I have added the 'crypto' tag.
- [ ] I have reviewed my diff in github, line by line and removed
unexpected formatting changes, testing logs, or commented-out code.
- [ ] Every change is related to the PR description.
- [ ] I have
[linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
this pull request to relevant issues (if any exist).

sirasistant committed:
chore: Workspace integration for the noir private kernel (#2753)

kevaundray committed:
chore: Initial code to generate ts code from Noir ABI (#2750)

This is some rough code to generate typescript bindings for a given
abi.json


# Checklist:
Remove the checklist to signal you've completed it. Enable auto-merge if
the PR is ready to merge.
- [ ] If the pull request requires a cryptography review (e.g.
cryptographic algorithm implementations) I have added the 'crypto' tag.
- [ ] I have reviewed my diff in github, line by line and removed
unexpected formatting changes, testing logs, or commented-out code.
- [ ] Every change is related to the PR description.
- [ ] I have
[linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
this pull request to relevant issues (if any exist).

---------

Co-authored-by: sirasistant <sirasistant@gmail.com>

sirasistant committed:
feat: convert params to noir types (Noir private kernel)  (#2760)

Please provide a paragraph or two giving a summary of the change,
including relevant motivation and context.

# Checklist:
Remove the checklist to signal you've completed it. Enable auto-merge if
the PR is ready to merge.
- [ ] If the pull request requires a cryptography review (e.g.
cryptographic algorithm implementations) I have added the 'crypto' tag.
- [ ] I have reviewed my diff in github, line by line and removed
unexpected formatting changes, testing logs, or commented-out code.
- [ ] Every change is related to the PR description.
- [ ] I have
[linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
this pull request to relevant issues (if any exist).

sirasistant committed:
feat: Add return type parsing for noir private kernel circuit (#2762)

Please provide a paragraph or two giving a summary of the change,
including relevant motivation and context.

# Checklist:
Remove the checklist to signal you've completed it. Enable auto-merge if
the PR is ready to merge.
- [ ] If the pull request requires a cryptography review (e.g.
cryptographic algorithm implementations) I have added the 'crypto' tag.
- [ ] I have reviewed my diff in github, line by line and removed
unexpected formatting changes, testing logs, or commented-out code.
- [ ] Every change is related to the PR description.
- [ ] I have
[linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
this pull request to relevant issues (if any exist).

TomAFrench commented on yarn-project/noir-private-kernel/src/crates/private-kernel-lib/src/utils/uint256.nr:
```diff
@@ -0,0 +1,131 @@
+
+impl U256 {
+    fn from_bytes32(bytes : [u8;32]) -> U256 {
+        let high_0 = ((bytes[0] as u64) << 56)
+                | ((bytes[1] as u64) << 48)
+                | ((bytes[2] as u64) << 40)
+                | ((bytes[3] as u64) << 32)
+                | ((bytes[4] as u64) << 24)
+                | ((bytes[5] as u64) << 16)
+                | ((bytes[6] as u64) << 8)
+                |  (bytes[7] as u64);
+        
+        let high_1 = ((bytes[8] as u64) << 56)
+                | ((bytes[9] as u64) << 48)
+                | ((bytes[10] as u64) << 40)
+                | ((bytes[11] as u64) << 32)
+                | ((bytes[12] as u64) << 24)
+                | ((bytes[13] as u64) << 16)
+                | ((bytes[14] as u64) << 8)
+                |  (bytes[15] as u64);
+        
+        let low_0 = ((bytes[16] as u64) << 56)
+                | ((bytes[17] as u64) << 48)
+                | ((bytes[18] as u64) << 40)
+                | ((bytes[19] as u64) << 32)
+                | ((bytes[20] as u64) << 24)
+                | ((bytes[21] as u64) << 16)
+                | ((bytes[22] as u64) << 8)
+                |  (bytes[23] as u64);
+        
+        let low_1 = ((bytes[24] as u64) << 56)
+                | ((bytes[25] as u64) << 48)
+                | ((bytes[26] as u64) << 40)
+                | ((bytes[27] as u64) << 32)
+                | ((bytes[28] as u64) << 24)
+                | ((bytes[29] as u64) << 16)
+                | ((bytes[30] as u64) << 8)
+                |  (bytes[31] as u64);
```
Note for later that we can reduce constraints by adding rather than ORing here.

kevaundray committed:
chore: Use Noir javascript packages to execute the private kernel circuit init using the typescript generated types (#2763)

kevaundray committed:
chore: enable noir private kernel circuit init (#2781)

kevaundray commented on yarn-project/circuits.js/src/structs/index.ts:
```diff
@@ -35,3 +35,10 @@ export * from '@aztec/foundation/eth-address';
 export * from '@aztec/foundation/aztec-address';
 export { FunctionSelector } from '@aztec/foundation/abi';
+
+// TODO(Kev): This is only exported so that privateKernelInit in noir-private-kernel
+// does not need to manually initialize its instance.
+// This is not great as we are exporting from tests.
+// Its okay for now and before merging into master, we should
+// remove this line.
```
Relink comment https://github.com/AztecProtocol/aztec-packages/pull/2763#discussion_r1353737765

kevaundray replied:
should we keep this re-export?

sirasistant replied:
It was removed :D

sirasistant committed:
fix: adding missing pieces for PKC in noir (#2782)

Please provide a paragraph or two giving a summary of the change,
including relevant motivation and context.

# Checklist:
Remove the checklist to signal you've completed it. Enable auto-merge if
the PR is ready to merge.
- [ ] If the pull request requires a cryptography review (e.g.
cryptographic algorithm implementations) I have added the 'crypto' tag.
- [ ] I have reviewed my diff in github, line by line and removed
unexpected formatting changes, testing logs, or commented-out code.
- [ ] Every change is related to the PR description.
- [ ] I have
[linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
this pull request to relevant issues (if any exist).

---------

Co-authored-by: kevaundray <kevtheappdev@gmail.com>

kevaundray committed:
make init have valid inputs

kevaundray committed:
add read_request_root_from_siblings wrapper method

kevaundray committed:
BROKEN BROKEN NOIR BUG:

Running `yarn noir:build && yarn test` on the noir-private-kernel-lib on this commit will give you a failure with the message `    Cannot satisfy constraint`.

The problem is `function_tree_root_from_siblings` which calls `root_from_sibling_path` which fails on L32, the indexing of `indices[i]`.

It is not an index out of bounds error, and seems to only happen with this test

kevaundray committed:
BUG BUG BUG SIMPLIFIED

kevaundray committed:
WORKAROUND FOR BUG NOTED IN PREVIOUS COMMIT

kevaundray committed:
update comment about reexporting tests

kevaundray committed:
Add a from_u32 method for FunctionSelector

kevaundray committed:
add interop tests

kevaundray committed:
uncomment code as bug has been committed

kevaundray committed:
add l2_to_l1_hash interop tests

To run the cpp tests, type the following into your shell:

- cd ~/aztec-packages/circuits/cpp
- cmake --preset default
- cmake --build --preset default
- ./build/bin/aztec3_utils_tests --gtest_filter=hash_tests.*

kevaundray committed:
chore: Change private-kernel init to use BoundedVec (#2800)

Co-authored-by: sirasistant <sirasistant@gmail.com>

sirasistant committed:
fix: mismatch in hashing for block hash

sirasistant committed:
fix: comment out membership checks and fixed the remaining issues (#2915)

sirasistant committed:
Merge branch 'master' into feature_branch/private-kernel

sirasistant committed:
fix: update types for pending rr

sirasistant committed:
feat: inner private kernel in noir (#2931)

Please provide a paragraph or two giving a summary of the change,
including relevant motivation and context.

# Checklist:
Remove the checklist to signal you've completed it. Enable auto-merge if
the PR is ready to merge.
- [ ] If the pull request requires a cryptography review (e.g.
cryptographic algorithm implementations) I have added the 'crypto' tag.
- [ ] I have reviewed my diff in github, line by line and removed
unexpected formatting changes, testing logs, or commented-out code.
- [ ] Every change is related to the PR description.
- [ ] I have
[linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
this pull request to relevant issues (if any exist).

sirasistant committed:
Merge branch 'master' into feature_branch/private-kernel

sirasistant committed:
fix: rename private data tree after merge

sirasistant committed:
feat: Ordering circuit in noir (#2979)

Please provide a paragraph or two giving a summary of the change,
including relevant motivation and context.

# Checklist:
Remove the checklist to signal you've completed it. Enable auto-merge if
the PR is ready to merge.
- [ ] If the pull request requires a cryptography review (e.g.
cryptographic algorithm implementations) I have added the 'crypto' tag.
- [ ] I have reviewed my diff in github, line by line and removed
unexpected formatting changes, testing logs, or commented-out code.
- [ ] Every change is related to the PR description.
- [ ] I have
[linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
this pull request to relevant issues (if any exist).

TomAFrench commented on yarn-project/noir-private-kernel/src/crates/private-kernel-lib/src/hash.nr:
```diff
@@ -0,0 +1,259 @@
+    node
+}
+
+fn u4_to_le_bit_array(value : u4) -> [u1;4] {
+    let mut le_bit_array : [u1; 4] = [0; 4];
+    for i in 0..4 {
+        le_bit_array[i] = ((value >> i) & 1) as u1;
+    }
+    le_bit_array
+}
+fn u16_to_le_bit_array(value : u16) -> [u1;16] {
+    let mut le_bit_array : [u1; 16] = [0; 16];
+    for i in 0..16 {
+        le_bit_array[i] = ((value >> i) & 1) as u1;
+    }
+    le_bit_array
+}
+fn u32_to_le_bit_array(value : u32) -> [u1;32] {
+    let mut le_bit_array : [u1; 32] = [0; 32];
+    for i in 0..32 {
+        le_bit_array[i] = ((value >> i) & 1) as u1;
+    }
+    le_bit_array
+}
```
```suggestion
fn u4_to_le_bit_array(value : u4) -> [u1;4] {
    let mut le_bit_array : [u1; 4] = [0; 4];
    for i in 0..4 {
        le_bit_array[i] = (value >> i) as u1;
    }
    le_bit_array
}
fn u16_to_le_bit_array(value : u16) -> [u1;16] {
    let mut le_bit_array : [u1; 16] = [0; 16];
    for i in 0..16 {
        le_bit_array[i] = (value >> i) as u1;
    }
    le_bit_array
}
fn u32_to_le_bit_array(value : u32) -> [u1;32] {
    let mut le_bit_array : [u1; 32] = [0; 32];
    for i in 0..32 {
        le_bit_array[i] = (value >> i) as u1;
    }
    le_bit_array
}
```

Casting to a `u1` performs the truncation we're doing with the `& 1` implicitly so we can remove it. This takes us from 272306 to 235186 constraints.

sirasistant committed:
Merge branch 'master' into feature_branch/private-kernel

sirasistant committed:
Merge branch 'master' into feature_branch/private-kernel

sirasistant committed:
feat: update to pedersen hash

sirasistant committed:
test: use snapshots

sirasistant committed:
test: bump timeouts

sirasistant committed:
test: fix block building test

sirasistant committed:
test: bump bench timeout further

sirasistant committed:
fix: use pedersen for addr nullifier (#3027)

sirasistant committed:
feat: uncomment read request membership check

sirasistant committed:
feat: fix contract membership check (#3034)

sirasistant committed:
Merge branch 'master' into feature_branch/private-kernel

sirasistant committed:
refactor: removed functions for bit decomposition

sirasistant committed:
test: add smoke test for noir PKC

sirasistant committed:
Merge branch 'master' into feature_branch/private-kernel

sirasistant committed:
restore cpp kernel circuits

sirasistant committed:
chore: mock vks also in cpp

sirasistant committed:
fix: revert vk hash changes

sirasistant committed:
refactor: cleaned up some todos and unused fns

sirasistant committed:
chore: undo boxes changes

sirasistant committed:
Merge branch 'master' into feature_branch/private-kernel

sirasistant reviewed (approved):
The noir private kernel has been removed from the PXE so this PR only adds the code and some basic smoke tests to make sure that we don't make a breaking change

sirasistant committed:
Merge branch 'master' into feature_branch/private-kernel

sirasistant committed:
style: fix formatting from master

sirasistant committed:
Merge branch 'master' into feature_branch/private-kernel

dan-aztec commented on yarn-project/noir-private-kernel/README.md:
```diff
@@ -0,0 +1 @@
+# Noir Private Kernel
```
sir!!
